public/js/test/mochitest/browser_dbg-sources.js

```diff
@@ -0,0 +1,58 @@
+/* Any copyright is dedicated to the Public Domain.
+ * http://creativecommons.org/publicdomain/zero/1.0/ */
+
+// Tests that the source tree works.
+
+function* waitForSourceCount(dbg, i) {
+  // We are forced to wait until the DOM nodes appear because the
+  // source tree batches its rendering.
+  yield waitUntil(() => {
+    return findAllElements(dbg, "sourceNodes").length === i;
+  });
+}
+
+add_task(function* () {
+  const dbg = yield initDebugger("doc-sources.html");
+  const { selectors: { getSelectedSource }, getState } = dbg;
+
+  // Expand nodes and make sure more sources appear.
+  is(findAllElements(dbg, "sourceNodes").length, 2);
+
+  clickElement(dbg, "sourceArrow", 2);
+  is(findAllElements(dbg, "sourceNodes").length, 7);
+
+  clickElement(dbg, "sourceArrow", 3);
+  is(findAllElements(dbg, "sourceNodes").length, 8);
+
+  // Select a source.
+  ok(!findElementWithSelector(dbg, ".sources-list .focused"),
+     "Source is not focused");
+  const selected = waitForDispatch(dbg, "SELECT_SOURCE");
+  clickElement(dbg, "sourceNode", 4);
+  yield selected;
```

Comment:
I think you meant `yield waitForDispatch(dbg, "SELECT_SOURCE");` in your original code ;)

I don't quite follow because I'm not dispatching anything here, I'm waiting for a dispatch. Oh, maybe you mean that if it is an async action, it would work to do something and then `waitForDispatch` right after because we don't care about the `"start"` action; our `waitForDispatch` will resolve with the `"done"` action which is guaranteed to be asynchronous. That's how it works most of the time.

`SELECT_SOURCE` is not an async action though. Not all `dispatch` calls return a promise. Only ones that fire an action creator that *themselves* return a promise do. `SELECT_SOURCE` is not async and is dispatched immediately in the `selectSource` action, which is fired synchronously when `clickElement` is called.

So we must have a handler for it waiting before `clickElement` is called. Otherwise, once `clickElement` returns the action is already dispatched and we've missed it.

This is a good thing to understand which is why I wanted to explain more. Once this is merged, feel free to try changing the code and watching it.